# Importing libraries

In [2]:
#Import libraries 
import pandas as pd
import numpy as np
import os

In [3]:
#Import data
path = r'/Users/buketoztekin/Documents/Instacart Basket Analysis/'
df_orders_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

# Checking for mixed type data

In [4]:
# Checking for mixed type data in df_orders_wrangled
for col in df_orders_wrangled.columns.tolist():
    weird = (df_orders_wrangled[col].map(type) != type(df_orders_wrangled[col].iloc[0])).any()
    if weird:
        print(col)

Unnamed: 0
order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [5]:
print(df_orders_wrangled.dtypes)

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object


In [6]:
# Convert column's data to string
df_orders_wrangled['order_id'] = df_orders_wrangled['order_id'].astype('str')
df_orders_wrangled['user_id'] = df_orders_wrangled['user_id'].astype('str')
df_orders_wrangled['order_number'] = df_orders_wrangled['order_number'].astype('str')
df_orders_wrangled['order_hour_of_day'] = df_orders_wrangled['order_hour_of_day'].astype('int64')
df_orders_wrangled['orders_day_of_week'] = df_orders_wrangled['orders_day_of_week'].astype('int64')
df_orders_wrangled['days_since_prior_order'] = df_orders_wrangled['days_since_prior_order'].astype('float64')

In [7]:
# Checking for mixed type data in df_orders_wrangled
for col in df_orders_wrangled.columns.tolist():
    weird = (df_orders_wrangled[col].map(type) != type(df_orders_wrangled[col].iloc[0])).any()
    if weird:
        print(col)

Unnamed: 0
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [8]:
# Checking for mixed type data in df_products
for col in df_products.columns.tolist():
    weird = (df_products[col].map(type) != type(df_products[col].iloc[0])).any()
    if weird:
        print(col)

product_id
product_name
aisle_id
department_id
prices


# Missing Values

In [9]:
#Finding missing values in df_products
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# Create subset of the dataframe that contains the nulls
df_nan = df_products[df_products['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

since string values (such as product_name) can't be imputed, we have to remove the missing data.

In [12]:
#Checking for row number
df_products.shape

(49693, 5)

In [13]:
#Creating a new dataframe without the missing values
df_products_clean = df_products[df_products['product_name'].isnull() == False]

In [14]:
#Checking for new row number
df_products_clean.shape

(49677, 5)

# Finding Duplicates

In [15]:
#Finding duplicates in the products_clean
df_dups = df_products_clean[df_products_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_products_clean.shape

(49677, 5)

In [18]:
#Creating a new dataframe without the duplicates
df_products_clean_no_dups = df_products_clean.drop_duplicates()

In [19]:
df_products_clean_no_dups.shape

(49672, 5)

# Exporting changes

In [20]:
df_products_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

## Step 2 Describing df_orders

In [21]:
df_orders.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Order number should be between 1-100
Orders day of the week is between 0-6 meaning that it can be any day of the week, starting from 0 (asumingly on Sunday)
Order hour of the day is between 0-23, representing 24 hours of the day.
''Days since prior order'' seems to have missing values. It is between 0 to 30 meaning that more than one orders can be place at the same day and also the maximum is capped at 30 days.

## Step 3 Checking for mixed type data

In [22]:
# Checking for mixed type data in df_orders
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].map(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col)

There are no mixed data types in the orders dataframe.

## Step 4

In [23]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


## Step 5 - Finding the missing values

In [24]:
#Finding missing values in df_orders
df_orders.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in days_since_prior_order column.

## Step 6 - Addressing the missing values

In [25]:
#Creating a subset of the missing values
df_orders_nan = df_orders[df_orders['days_since_prior_order'].isnull() == True]

In [26]:
df_orders_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

From the above table, we know that the mean of dats_since_prior_order is 11.
We should also check the median and decide which one we will use the impute the data.

In [27]:
df_orders['days_since_prior_order'].median()

7.0

The mean(11) is higher than the median (7) which indicates that the data is right-skewed. Therefore, it is better to use the median which is not influenced by the outliers. 

In [36]:
#Creating a copy of df_orders
df_orders_clean = df_orders.copy()

In [37]:
#Creating a new dataframe without the missing values
df_orders_clean['days_since_prior_order'] = df_orders_clean['days_since_prior_order'].fillna(7)

In [38]:
#Checking for missing values in df_orders_clean
df_orders_clean.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [41]:
df_orders_clean.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     7.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Step 7 - Finding duplicates

In [42]:
#Checking for duplicates
df_orders_clean_dups = df_orders_clean[df_orders_clean.duplicated()]

In [43]:
df_orders_clean_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates in the dataframe.

## Step 8 - Exporting data

In [44]:
df_orders_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))